fetcher/fetcher.go

```diff
@@ -0,0 +1,228 @@
+// a value is only fetched from disk once. Subsequent
+// requests can be retrieved from cache.
+type Fetcher struct {
+	im state.Immutable
+
+	keysToFetch map[string]*keyData        // Number of txns waiting for a key
+	txnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
+	keyLock     sync.RWMutex
+	txnLock     sync.Mutex
+
+	stopOnce  sync.Once
+	stop      chan struct{}
+	err       error
+	fetchable chan *task
+	l         sync.RWMutex
+	wg        sync.WaitGroup
+
+	completed int
+	numTxs    int
+}
+
+// Data to insert into the cache
+type fetchData struct {
+	v      []byte
+	exists bool
+	chunks uint16
+}
+
+// Data pertaining to whether a key was fetched or which txns are waiting
+type keyData struct {
+	cache *fetchData
+	queue []ids.ID
+}
+
+// Holds the information that a worker needs to fetch values
+type task struct {
+	ctx context.Context
+	id  ids.ID
+	key string
+}
+
+// New creates a new [Fetcher]
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		keysToFetch: make(map[string]*keyData),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task, numTxs),
+		stop:        make(chan struct{}),
+		numTxs:      numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.wg.Add(1)
+		go f.runWorker()
+	}
+	return f
+}
+
+// Workers fetch individual keys
+func (f *Fetcher) runWorker() {
+	defer f.wg.Done()
+
+	for {
+		select {
+		case t, ok := <-f.fetchable:
+			if !ok {
+				return
+			}
+
+			// Allows concurrent reads to cache.
+			if ok := f.shouldFetch(t); !ok {
+				continue
+			}
+
+			// Fetch from disk once if key isn't already in cache
+			v, err := f.im.GetValue(t.ctx, []byte(t.key))
+			if errors.Is(err, database.ErrNotFound) {
+				f.update(t.key, nil, false, 0)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			numChunks, ok := keys.NumChunks(v)
+			if !ok {
+				f.stopOnce.Do(func() {
+					f.err = ErrInvalidKeyValue
+					close(f.stop)
+				})
+				return
+			}
+			f.update(t.key, v, true, numChunks)
+		case <-f.stop:
+			return
+		}
+	}
+}
+
+// Checks if a key is in the cache. If it's not in cache, check
+// if we were the first to request the key.
+func (f *Fetcher) shouldFetch(t *task) bool {
+	f.keyLock.RLock()
+	defer f.keyLock.RUnlock()
+
+	if f.keysToFetch[t.key].cache != nil {
+		return false
+	}
+	return f.keysToFetch[t.key].queue[0] == t.id
+}
+
+// Updates the cache and dependencies
+func (f *Fetcher) update(k string, v []byte, exists bool, chunks uint16) {
+	f.keyLock.Lock()
+	// Puts a key that was fetched from disk into cache
+	f.keysToFetch[k].cache = &fetchData{v, exists, chunks}
+	queue := f.keysToFetch[k].queue
+	f.keysToFetch[k].queue = nil
+	f.keyLock.Unlock()
+
+	f.txnLock.Lock()
+	for _, id := range queue {
+		f.txnsToFetch[id].Done() // Notify all other txs
```

Comment:
Ah, this is probably a race between `txnsToFetch` and `f.keysToFetch[k].queue = append(f.keysToFetch[k].queue, txID)`. This is why I suggested you could unify things behind a single lock instead of trying to use 2.